### PR TITLE
gh-111784: Fix two segfaults

### PR DESCRIPTION
**Pull request by Eclips4**
This PR contains two fixes.
First fix resolve situation when `pyexpat` module (which contains `expat_CAPI` capsule) deallocates before `_elementtree`, so we need to hold a strong reference to `pyexpat` module to.

Second fix resolve situation when module state is deallocated before deallocation of `XMLParser` instances, which uses module state to clear some stuff.

(I'll write a `NEWS` entry later, if it is necessary)
* Issue: gh-111784

### CODE REVIEW

**Comment by Eclips4** [comment]
@mgorny Can you confirm that problem is resolved now?

**Comment by Eclips4** [comment]
Also, I check it for reference leaks:
```python
./python.exe -m test -R 3:3 test_xml_etree_c
Using random seed: 1167552768
0:00:00 load avg: 2.77 Run 1 test sequentially
0:00:00 load avg: 2.77 [1/1] test_xml_etree_c
beginning 6 repetitions
123456
......

== Tests result: SUCCESS ==

1 test OK.

Total duration: 2.7 sec
Total tests: run=208 skipped=6
Total test files: run=1/1
Result: SUCCESS
```

**Comment by mgorny** [comment]
Thanks! I can confirm that with this patch, slixmpp's test suite passes without segfaults on top of Python 3.12.

**Review by serhiy-storchaka — APPROVED** [review]
LGTM.

Yes, add a NEWS entry, please.

**Comment by Eclips4** [comment]
> LGTM.
> 
> Yes, add a NEWS entry, please.

Done.

**Review by serhiy-storchaka — APPROVED** [review]

File: Misc/NEWS.d/next/Library/2023-12-23-13-10-42.gh-issue-111784.Nb4L1j.rst [hunk 1]
```diff
@@ -0,0 +1,4 @@
+Fix first segfault during deallocation of ``_elementtree.XMLParser`` instances by keeping strong reference
+to ``pyexpat`` module in module state for capsule lifetime.
+Fix second segfault which happens in the same deallocation process  by keeping strong reference
+to ``_elementtree`` module in ``XMLParser`` structure for ``_elementtree`` module lifetime.
```
Comment (serhiy-storchaka) [inline Misc/NEWS.d/next/Library/2023-12-23-13-10-42.gh-issue-111784.Nb4L1j.rst]: A newline is need at the end of the last line. Also, some context was lost.
```suggestion
Fix segfaults in the ``_elementtree`` module.
Fix first segfault during deallocation of ``_elementtree.XMLParser`` instances by keeping strong reference
to ``pyexpat`` module in module state for capsule lifetime.
Fix second segfault which happens in the same deallocation process by keeping strong reference
to ``_elementtree`` module in ``XMLParser`` structure for ``_elementtree`` module lifetime.
```

**Comment by Eclips4** [comment]
Thanks Serhiy for your review!

**Pull request merged**